### PR TITLE
Change the encoding to JS encoder

### LINKED ISSUE #89: Empty buffer as values
add support so that test: ``should support retrieving empty buffers as values`` passes
https://github.com/scylladb-zpp-2024-javascript-driver/scylladb-javascript-driver/blob/bad25f7a837a434efa6aa3a0c50157c18e62ce46/test/integration/not-supported/client-execute-tests.js#L131-L168

### LINKED ISSUE #167: Verify consistency between behavior of new and old UDT
For now during wrapping UDT into expected type, unused fields will be omitted and no error will be thrown.
Thats need to be verified with DataStax logic.

### LINKED ISSUE #200: Impoprer handling of unset
Somehow, this missed also my review.
https://github.com/scylladb-zpp-2024-javascript-driver/scylladb-javascript-driver/blob/1f87b2f26d42650f32b84fe5e83c452579520884/lib/types/cql-utils.js#L323-L334

Fix this and maybe add regression tests for that

### LINKED ISSUE #215: TypeGuessing for Tuple and UDT

### LINKED ISSUE #233: Unsets inside collections
Tuples and UDTs can have unsets inside. 
Other collections like lists or sets should be verified if they can or cannot have unsets inside.

in `cqlUtils` those types wich can have unsets inside should use `getWrapped` instead of `wrapBasedOnType`

### LINKED ISSUE #245: Missing column metadata information
Missing column metadata
https://github.com/scylladb-zpp-2024-javascript-driver/scylladb-javascript-driver/blob/main/lib/types/results-wrapper.js#L126-L133

See https://github.com/scylladb-zpp-2024-javascript-driver/scylladb-javascript-driver/pull/168#discussion_r2280960128_

### PR DESCRIPTION
**Pull request by adespawn**
The goal of this PR is to change the current encoding approach (as described in #257) to Approach 1.

~~Currently, UDTs are not supported because the proper type information conversion has not yet been implemented.~~ UDTs are supported in this PR.

This PR also introduces a change to the prepared execute API of the Rust wrapper. It replaces PreparedStatementWrapper with strings when executing prepared statements during query execution at the endpoint. With this change, we only return expected types for prepared statements, instead of the whole PreparedStatementWrapper object from Rust.
This change was introduced for performance reasons, as it results in an improvement when executing queries
through executeConcurrent, and does not slow down other endpoints. 

### Testing

This PR was tested locally with Cassandra integration tests, on top of the CI running the Scylla integration tests (see #244)

### Encoding issues

#257

By changing the encoder, the following issues would be fixed:

fixes #89
fixes #167 
fixes #200 
fixes #233 
fixes #215 (Type guessing is done inside encoder)

With proper type conversion, this PR fixes #245

### Performance

While the core part of this PR (replacing encoders) yields a decent speedup, it is only with the other optimisation that we reach the same speed as the test implementation of the other approach.

Performance of the baseline (main):

```bash
➜ sudo perf stat node ./benchmark/logic/concurrent_insert.js scylladb-javascript-driver 2000000

 Performance counter stats for 'node ./benchmark/logic/concurrent_insert.js scylladb-javascript-driver 2000000':

         97,246.47 msec task-clock                       #    2.827 CPUs utilized             
         3,235,539      context-switches                 #   33.272 K/sec                     
           399,293      cpu-migrations                   #    4.106 K/sec                     
           444,992      page-faults                      #    4.576 K/sec                     
   158,132,408,757      cpu_atom/instructions/           #    0.70  insn per cycle              (45.09%)
   441,422,538,978      cpu_core/instructions/           #    1.13  insn per cycle              (54.91%)
   225,794,559,768      cpu_atom/cycles/                 #    2.322 GHz                         (45.09%)
   391,037,324,170      cpu_core/cycles/                 #    4.021 GHz                         (54.91%)
    30,086,164,897      cpu_atom/branches/               #  309.381 M/sec                       (45.09%)
    90,474,152,804      cpu_core/branches/               #  930.359 M/sec                       (54.91%)
       486,036,509      cpu_atom/branch-misses/          #    1.62% of all branches             (45.09%)
       658,875,141      cpu_core/branch-misses/          #    0.73% of all branches             (54.91%)
             TopdownL1 (cpu_atom)                 #     30.0 %  tma_backend_bound        (45.09%)
                                                  #     41.7 %  tma_backend_bound      
                                                  #      6.1 %  tma_bad_speculation    
                                                  #     32.2 %  tma_frontend_bound     
                                                  #     20.1 %  tma_retiring             (54.91%)
             TopdownL1 (cpu_atom)                 #     14.9 %  tma_retiring             (45.09%)
             TopdownL1 (cpu_atom)                 #      7.4 %  tma_bad_speculation    
                                                  #     47.7 %  tma_frontend_bound       (45.09%)

      34.395850085 seconds time elapsed

      57.638440000 seconds user
      33.994388000 seconds sys
```

Performance for this implementation:

```bash
➜ sudo perf stat node ./benchmark/logic/concurrent_insert.js scylladb-javascript-driver 2000000

 Performance counter stats for 'node ./benchmark/logic/concurrent_insert.js scylladb-javascript-driver 2000000':

         89,723.35 msec task-clock                       #    3.182 CPUs utilized             
         3,258,107      context-switches                 #   36.313 K/sec                     
           349,166      cpu-migrations                   #    3.892 K/sec                     
           418,220      page-faults                      #    4.661 K/sec                     
   144,066,525,376      cpu_atom/instructions/           #    0.69  insn per cycle              (48.61%)
   354,393,050,737      cpu_core/instructions/           #    1.00  insn per cycle              (51.39%)
   207,875,282,834      cpu_atom/cycles/                 #    2.317 GHz                         (48.61%)
   352,819,743,579      cpu_core/cycles/                 #    3.932 GHz                         (51.39%)
    27,387,146,881      cpu_atom/branches/               #  305.240 M/sec                       (48.61%)
    73,594,614,458      cpu_core/branches/               #  820.239 M/sec                       (51.39%)
       403,763,266      cpu_atom/branch-misses/          #    1.47% of all branches             (48.61%)
       613,901,763      cpu_core/branch-misses/          #    0.83% of all branches             (51.39%)
             TopdownL1 (cpu_atom)                 #     31.9 %  tma_backend_bound        (48.61%)
                                                  #     40.3 %  tma_backend_bound      
                                                  #      5.9 %  tma_bad_speculation    
                                                  #     35.7 %  tma_frontend_bound     
                                                  #     18.1 %  tma_retiring             (51.39%)
             TopdownL1 (cpu_atom)                 #     14.9 %  tma_retiring             (48.61%)
             TopdownL1 (cpu_atom)                 #      7.1 %  tma_bad_speculation    
                                                  #     46.1 %  tma_frontend_bound       (48.61%)

      28.197915381 seconds time elapsed

      50.147893000 seconds user
      34.252507000 seconds sys

```

This + Strings instead of Prepared Statement Wrappers:

```bash
➜ sudo perf stat node ./benchmark/logic/concurrent_insert.js scylladb-javascript-driver 2000000

 Performance counter stats for 'node ./benchmark/logic/concurrent_insert.js scylladb-javascript-driver 2000000':

         81,705.69 msec task-clock                       #    3.333 CPUs utilized             
         2,540,136      context-switches                 #   31.089 K/sec                     
           274,768      cpu-migrations                   #    3.363 K/sec                     
           418,423      page-faults                      #    5.121 K/sec                     
   138,566,100,653      cpu_atom/instructions/           #    0.71  insn per cycle              (46.94%)
   344,678,249,478      cpu_core/instructions/           #    1.04  insn per cycle              (53.06%)
   195,016,125,911      cpu_atom/cycles/                 #    2.387 GHz                         (46.94%)
   330,478,854,641      cpu_core/cycles/                 #    4.045 GHz                         (53.06%)
    26,103,495,597      cpu_atom/branches/               #  319.482 M/sec                       (46.94%)
    71,403,457,264      cpu_core/branches/               #  873.910 M/sec                       (53.06%)
       357,153,938      cpu_atom/branch-misses/          #    1.37% of all branches             (46.94%)
       522,466,210      cpu_core/branch-misses/          #    0.73% of all branches             (53.06%)
             TopdownL1 (cpu_atom)                 #     33.6 %  tma_backend_bound        (46.94%)
                                                  #     43.3 %  tma_backend_bound      
                                                  #      5.6 %  tma_bad_speculation    
                                                  #     30.5 %  tma_frontend_bound     
                                                  #     20.6 %  tma_retiring             (53.06%)
             TopdownL1 (cpu_atom)                 #     15.2 %  tma_retiring             (46.94%)
             TopdownL1 (cpu_atom)                 #      6.7 %  tma_bad_speculation    
                                                  #     44.5 %  tma_frontend_bound       (46.94%)

      24.514119238 seconds time elapsed

      48.662264000 seconds user
      28.094413000 seconds sys


```

Performance of the alternative implementation (see Approach 2 in #257)

```bash
➜ sudo perf stat node ./benchmark/logic/concurrent_insert.js scylladb-javascript-driver 2000000

 Performance counter stats for 'node ./benchmark/logic/concurrent_insert.js scylladb-javascript-driver 2000000':

         78,012.38 msec task-clock                       #    3.161 CPUs utilized             
         2,434,806      context-switches                 #   31.211 K/sec                     
           266,106      cpu-migrations                   #    3.411 K/sec                     
           419,842      page-faults                      #    5.382 K/sec                     
   134,718,558,269      cpu_atom/instructions/           #    0.73  insn per cycle              (46.18%)
   336,355,949,157      cpu_core/instructions/           #    1.08  insn per cycle              (53.82%)
   185,637,195,079      cpu_atom/cycles/                 #    2.380 GHz                         (46.18%)
   312,091,083,187      cpu_core/cycles/                 #    4.001 GHz                         (53.82%)
    25,320,442,441      cpu_atom/branches/               #  324.570 M/sec                       (46.18%)
    69,254,709,870      cpu_core/branches/               #  887.740 M/sec                       (53.82%)
       343,029,928      cpu_atom/branch-misses/          #    1.35% of all branches             (46.18%)
       519,561,406      cpu_core/branch-misses/          #    0.75% of all branches             (53.82%)
             TopdownL1 (cpu_atom)                 #     32.7 %  tma_backend_bound        (46.18%)
                                                  #     38.5 %  tma_backend_bound      
                                                  #      6.5 %  tma_bad_speculation    
                                                  #     35.5 %  tma_frontend_bound     
                                                  #     19.4 %  tma_retiring             (53.82%)
             TopdownL1 (cpu_atom)                 #     15.4 %  tma_retiring             (46.18%)
             TopdownL1 (cpu_atom)                 #      6.8 %  tma_bad_speculation    
                                                  #     45.1 %  tma_frontend_bound       (46.18%)

      24.679384241 seconds time elapsed

      45.424897000 seconds user
      27.732913000 seconds sys

```

Performance of the DataStax driver:

```bash
➜ sudo perf stat node ./benchmark/logic/concurrent_insert.js cassandra-driver 2000000

 Performance counter stats for 'node ./benchmark/logic/concurrent_insert.js cassandra-driver 2000000':

         26,551.61 msec task-clock                       #    0.959 CPUs utilized             
            27,071      context-switches                 #    1.020 K/sec                     
             3,349      cpu-migrations                   #  126.132 /sec                      
           449,716      page-faults                      #   16.937 K/sec                     
    88,366,241,047      cpu_atom/instructions/           #    1.20  insn per cycle              (9.14%)
   189,619,403,347      cpu_core/instructions/           #    1.62  insn per cycle              (90.86%)
    73,571,810,673      cpu_atom/cycles/                 #    2.771 GHz                         (9.14%)
   116,797,644,248      cpu_core/cycles/                 #    4.399 GHz                         (90.86%)
    18,830,161,128      cpu_atom/branches/               #  709.191 M/sec                       (9.14%)
    40,092,228,477      cpu_core/branches/               #    1.510 G/sec                       (90.86%)
       158,303,368      cpu_atom/branch-misses/          #    0.84% of all branches             (9.14%)
       147,831,358      cpu_core/branch-misses/          #    0.37% of all branches             (90.86%)
             TopdownL1 (cpu_atom)                 #     46.9 %  tma_backend_bound        (9.14%)
                                                  #     42.6 %  tma_backend_bound      
                                                  #      6.4 %  tma_bad_speculation    
                                                  #     23.7 %  tma_frontend_bound     
                                                  #     27.3 %  tma_retiring             (90.86%)
             TopdownL1 (cpu_atom)                 #     21.1 %  tma_retiring             (9.14%)
             TopdownL1 (cpu_atom)                 #      7.4 %  tma_bad_speculation    
                                                  #     24.6 %  tma_frontend_bound       (9.14%)

      27.680081974 seconds time elapsed

      23.598783000 seconds user
       2.296997000 seconds sys
```

Regular insert, this with Prepared Statement Wrapper

```bash
➜ sudo perf stat node ./benchmark/logic/insert.js scylladb-javascript-driver 200000

 Performance counter stats for 'node ./benchmark/logic/insert.js scylladb-javascript-driver 200000':

         17,006.43 msec task-clock                       #    0.871 CPUs utilized             
         1,424,966      context-switches                 #   83.790 K/sec                     
            10,045      cpu-migrations                   #  590.659 /sec                      
            10,575      page-faults                      #  621.824 /sec                      
    32,552,691,527      cpu_atom/instructions/           #    0.80  insn per cycle              (11.45%)
    67,934,051,811      cpu_core/instructions/           #    1.10  insn per cycle              (88.55%)
    40,840,442,735      cpu_atom/cycles/                 #    2.401 GHz                         (11.45%)
    61,553,896,677      cpu_core/cycles/                 #    3.619 GHz                         (88.55%)
     6,556,868,857      cpu_atom/branches/               #  385.552 M/sec                       (11.45%)
    13,958,890,195      cpu_core/branches/               #  820.801 M/sec                       (88.55%)
       109,830,162      cpu_atom/branch-misses/          #    1.68% of all branches             (11.45%)
       154,957,546      cpu_core/branch-misses/          #    1.11% of all branches             (88.55%)
             TopdownL1 (cpu_atom)                 #     25.3 %  tma_backend_bound        (11.45%)
                                                  #     31.7 %  tma_backend_bound      
                                                  #     11.0 %  tma_bad_speculation    
                                                  #     30.1 %  tma_frontend_bound     
                                                  #     27.3 %  tma_retiring             (88.55%)
             TopdownL1 (cpu_atom)                 #     15.9 %  tma_retiring             (11.45%)
             TopdownL1 (cpu_atom)                 #      8.6 %  tma_bad_speculation    
                                                  #     50.2 %  tma_frontend_bound       (11.45%)

      19.519492205 seconds time elapsed

       9.844989000 seconds user
       6.927643000 seconds sys

```

Regular insert, this with String

```bash
➜ sudo perf stat node ./benchmark/logic/insert.js scylladb-javascript-driver 200000 

 Performance counter stats for 'node ./benchmark/logic/insert.js scylladb-javascript-driver 200000':

         17,233.65 msec task-clock                       #    0.879 CPUs utilized             
         1,438,062      context-switches                 #   83.445 K/sec                     
             9,721      cpu-migrations                   #  564.071 /sec                      
            10,617      page-faults                      #  616.062 /sec                      
    32,687,188,606      cpu_atom/instructions/           #    0.81  insn per cycle              (9.57%)
    67,994,125,430      cpu_core/instructions/           #    1.09  insn per cycle              (90.43%)
    40,553,070,823      cpu_atom/cycles/                 #    2.353 GHz                         (9.57%)
    62,539,738,970      cpu_core/cycles/                 #    3.629 GHz                         (90.43%)
     6,551,270,505      cpu_atom/branches/               #  380.144 M/sec                       (9.57%)
    13,945,795,735      cpu_core/branches/               #  809.219 M/sec                       (90.43%)
        94,236,411      cpu_atom/branch-misses/          #    1.44% of all branches             (9.57%)
       160,288,721      cpu_core/branch-misses/          #    1.15% of all branches             (90.43%)
             TopdownL1 (cpu_atom)                 #     30.2 %  tma_backend_bound        (9.57%)
                                                  #     27.0 %  tma_backend_bound      
                                                  #     12.1 %  tma_bad_speculation    
                                                  #     36.3 %  tma_frontend_bound     
                                                  #     24.6 %  tma_retiring             (90.43%)
             TopdownL1 (cpu_atom)                 #     16.4 %  tma_retiring             (9.57%)
             TopdownL1 (cpu_atom)                 #      7.5 %  tma_bad_speculation    
                                                  #     45.9 %  tma_frontend_bound       (9.57%)

      19.611854331 seconds time elapsed

       9.939355000 seconds user
       7.052177000 seconds sys
```

### CODE REVIEW

File: lib/client.js [hunk 1]
```diff
@@ -81,6 +84,9 @@ class Client extends events.EventEmitter {
          * @type {ClientMetrics}
          */
         this.metrics = this.options.metrics;
+
+        // TODO: This field is currently hardcoded. Should be implemented properly
+        this.#encoder = new Encoder(0x04, this.options.encoding);
```
[the remark this answers is not in the document]
Comment (adespawn) [reply lib/client.js]: I wasn't able to find any documentation for that. Maybe I wasn't looking closely enough

File: src/types/encoded_data.rs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    cluster::metadata::ColumnType,
+    errors::SerializationError,
+    serialize::value::{BuiltinSerializationError, BuiltinSerializationErrorKind, SerializeValue},
+};
+
+enum EitherNullOrUnsetOrValue<T> {
+    Value(T),
+    Null,
+    Unset,
+}
```
[the remark this answers is not in the document]
Comment (adespawn) [reply src/types/encoded_data.rs]: I feel like having just one level of nested enums is cleaner, but your name suggestion is definitely better than the current one

File: src/types/encoded_data.rs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    unsafe fn from_napi_value(
+        env: napi::sys::napi_env,
+        napi_val: napi::sys::napi_value,
+    ) -> napi::Result<Self> {
+        let mut val_type: i32 = 0;
+        check_status!(
+            // Caller of this function ensures a valid pointer to napi env is provided
+            unsafe { sys::napi_typeof(env, napi_val, &mut val_type) },
+            "Failed to convert napi value into rust type `EncodedValuesWrapper`",
+        )?;
```
[the remark this answers is not in the document]
Comment (adespawn) [reply src/types/encoded_data.rs]: I've copied it from the implementation of ToNapiValue for Option<T>, so it may be doc hidden, or even without a documentation

File: test/unit/encoder-tests.js [hunk 4]
```diff
@@ -1248,7 +1248,7 @@ describe("encoder", function () {
             const encoder = new Encoder(4, {
                 encoding: { useUndefinedAsUnset: true },
             });
-            assert.strictEqual(encoder.encode(undefined), types.unset);
+            assert.strictEqual(encoder.encode(undefined), undefined);
```
[the remark this answers is not in the document]
Comment (adespawn) [reply test/unit/encoder-tests.js]: This is only internal representation of the unset

File: src/types/encoded_data.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (wprzytula) [reply src/types/encoded_data.rs]: OK, then feel free to choose whichever you prefer.

File: lib/client.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (wprzytula) [reply lib/client.js]: Indeed, there's no documentation. This is because CQL v4 is the only protocol version supported by ScyllaDB itself :P

File: test/unit/encoder-tests.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (wprzytula) [reply test/unit/encoder-tests.js]: OK, valid.

File: lib/types/cql-utils.js [hunk 5]
```diff
@@ -480,6 +480,29 @@ function convertHints(hints) {
...
+        data.info = [convertComplexType(fistSupport)];
+        if (secondSupport != null) {
+            data.info.push(convertComplexType(secondSupport));
+        }
+    } else if (otherTypes.length > 0) {
+        data.info = otherTypes.map((t) => convertComplexType(t));
+    }
+    return data;
+}
+
```
[the remark this answers is not in the document]
Comment (adespawn) [reply lib/types/cql-utils.js]: > Is this feasible? Is this worth it?

While possible, I'm not sure it's worth it as the user requires the same format for the metadata of the request.

File: lib/client.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (adespawn) [reply lib/client.js]: Sure.
https://github.com/scylladb/scylladb/issues/26574

File: src/types/encoded_data.rs [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+            sys::ValueType::napi_null => Ok(EncodedValuesWrapper {
+                inner: EitherNullOrUnsetOrValue::Null,
+            }),
+            _ => {
+                // Caller of this function ensures a valid pointer to napi env is provided
+                let v = unsafe { Uint8Array::from_napi_value(env, napi_val)? };
+                let z: &[u8] = &v;
+                Ok(EncodedValuesWrapper {
+                    inner: EitherNullOrUnsetOrValue::Value(z.to_vec()),
+                })
```
[the remark this answers is not in the document]
Comment (adespawn) [reply src/types/encoded_data.rs]: #116

File: lib/encoder.js [hunk 7]
```diff
@@ -1129,7 +1135,7 @@ function defineInstanceMembers() {
     this.encodeTuple = function (value, tupleInfo) {
         const parts = [];
         let totalLength = 0;
-        const length = Math.min(tupleInfo.length, value.length);
+        const length = Math.max(tupleInfo.length, value.length);
```
[the remark this answers is not in the document]
Comment (adespawn) [reply lib/encoder.js]: It turns out this causes test to fail due to decoding. This is either Rust driver bug, or the driver inserts invalid data into database, and the DB does not validate it. Examples for this bug can be found here:
https://github.com/scylladb/nodejs-rs-driver/commit/b0e8fc9c8be9351d627256bb415d22a48fb900f4

File: test/integration/supported/client-execute-prepared-tests.js [hunk 8]
```diff
@@ -466,13 +475,17 @@ describe("Client @SERVER_API", function () {
...
+                                value = value.multiply(
+                                    types.Integer.fromString("9999901443"),
+                                );
                                 if (n % 2 === 0) {
                                     // as a string also
                                     value = value.toString();
                                 }
+                                expectedRows[id] = arbitraryValueToBigInt(
+                                    value.toString(),
+                                );
```
[the remark this answers is not in the document]
Comment (adespawn) [reply test/integration/supported/client-execute-prepared-tests.js]: Added a comment explaining that

File: lib/encoder.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (wprzytula) [reply lib/encoder.js]: Let's open an issue about this.

[the remark this answers is not in the document]
Comment (adespawn) [reply lib/encoder.js]: https://github.com/scylladb/scylla-rust-driver/issues/1452

File: lib/types/cql-utils.js [hunk 9]
```diff
@@ -480,6 +480,29 @@ function convertHints(hints) {
...
+ * @param {rust.ComplexType} type
+ */
+function convertComplexType(type) {
+    let data = {
+        code: type.baseType.valueOf(),
+        info: null,
+    };
+    let fistSupport = type.getFirstSupportType();
+    let secondSupport = type.getSecondSupportType();
+    let otherTypes = type.getInnerTypes();
```
[the remark this answers is not in the document]
Comment (wprzytula) [reply lib/types/cql-utils.js]:
```suggestion
    let innerTypes = type.getInnerTypes();
```

**Review by wprzytula — CHANGES_REQUESTED** [review]

**Review by wprzytula — APPROVED** [review]
